android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboardDownloaderActivity.java

```diff
@@ -113,267 +118,280 @@ public void onCreate(Bundle savedInstanceState) {
       }
 
       title = customLanguage + "_" + customKeyboard;
-  } else {
-    // Download keyboard from cloud server
-    title = langName + ": " + kbName;
-  }
+    } else {
+      // Download keyboard from cloud server
+      title = langName + ": " + kbName;
+    }
 
-  DialogFragment dialog = new ConfirmDialogFragment();
-    args.putString(ConfirmDialogFragment.ARG_TITLE, title);
-    dialog.setArguments(args);
-    dialog.show(getFragmentManager(), "dialog");
-}
+    DialogFragment dialog = new ConfirmDialogFragment();
+      args.putString(ConfirmDialogFragment.ARG_TITLE, title);
+      dialog.setArguments(args);
+      dialog.show(getFragmentManager(), "dialog");
+  }
 
   /**
-   * Async task to download a Keyman keyboard from either Keyman cloud server or custom url
-   * @param context
-   * @param showProgressDialog
+   * Used by the <code>download</code> method to handle asynchronous downloading and evaluation of
+   * packages and keyboards.
    */
-    public static void download(final Context context, final boolean showProgressDialog) {
-
-    new AsyncTask<Void, Integer, Integer>() {
-      private ProgressDialog progressDialog;
-      //private String languageName = "";
-      //private String keyboardName = "";
-      private String kbVersion = "1.0";
-      private String kbIsCustom = isCustom ? "Y" : "N";
-      private String font = "";
-      private String oskFont = "";
-
-      @Override
-      protected void onPreExecute() {
-        super.onPreExecute();
-        if (showProgressDialog) {
-          progressDialog = new ProgressDialog(context);
-          progressDialog.setMessage("Downloading keyboard...");
-          progressDialog.setCancelable(false);
-          if (!((Activity) context).isFinishing()) {
-            progressDialog.show();
-          } else {
-            cancel(true);
-            progressDialog = null;
-          }
+  static class DownloadTask extends AsyncTask<Void, Integer, Integer> {
+    private ProgressDialog progressDialog;
+    private String kbVersion = "1.0";
+    private String kbIsCustom = isCustom ? "Y" : "N";
+    private String font = "";
+    private String oskFont = "";
+
+    private Context context;
+    private boolean showProgressDialog;
+
+    public DownloadTask(Context context, boolean showProgressDialog) {
+      this.context = context;
+      this.showProgressDialog = showProgressDialog;
+    }
+
+    @Override
+    protected void onPreExecute() {
+      super.onPreExecute();
+      if (showProgressDialog) {
+        progressDialog = new ProgressDialog(context);
+        progressDialog.setMessage("Downloading keyboard...");
+        progressDialog.setCancelable(false);
```

Comment:
I personally find progress dialogs that can't be cancelled unfriendly. Is there any easy way we can make it cancelable?

Reply:
It's easy to add the "Cancel" button, but it might be tricky to get the download activities to finish abruptly and gracefully. I'll save this for last.